Export attestations for local exporter

jedevc opened the pull request:
Implements the `local` exporter part of #3184.

The attestation extraction logic is moved to a separate utils package, so that other exporters can use it. Additionally, the bundling logic is slightly reworked, to add an explicit Unbundle step before Extracting. This ensures that the Extract method can perform a one-to-one translation from result.Attestations to intoto.Statements, which allows for deriving intermediate properties such as the Path.

The attestation files are exported similar to another platform ref, with file names equivalent to the original attestation files had (though this should be pretty easy to change if we'd prefer a separate output location specified by the user).

I'm not quite sure what to put for the `subject` field for intoto (unless there's a canonical way of calculating the hash of a directory that I'm not aware of?), so for now I've left those empty.

tonistiigi commented:
You can use this pkg https://github.com/moby/buildkit/compare/master...tonistiigi:buildkit:staticfs?expand=1 to add new static files(attestation data) to the transfer context and then merge it with the existing `FS` implementation.

tonistiigi commented on exporter/util/attestation/extract.go:
```diff
@@ -0,0 +1,260 @@
+package attestation
```
Packages that import solver, ImmutableRef etc. should not be under util. They can still be in a separate subpackage if it makes sense.

tonistiigi commented on exporter/local/export.go:
```diff
@@ -149,6 +152,41 @@ func (e *localExporterInstance) Export(ctx context.Context, inp *exporter.Source
+						return errors.Wrap(err, "failed to marshal attestation")
+					}
+
+					name := path.Base(attestations[i].Path)
```
I think there should be a way to set prefix for this name with attestation opts (usually binary name that is exported with `type=local`).

tonistiigi commented on client/client_test.go:
```diff
@@ -6942,110 +6942,153 @@ func testExportAttestations(t *testing.T, sb integration.Sandbox) {
+			require.Equal(t, "https://example.com/attestations2/v1.0", attest2.PredicateType)
+			require.Nil(t, attest2.Predicate)
+			subjects := []intoto.Subject{{
+				Name: "/attestation.json",
```
Not quite sure what is going on here. I think we discussed that subject should be all files (so "greeting"?). Hash is sha256 of content only, no metadata. Directories ignored.

jedevc replied:
Have updated, so that the `Self` subject will create a subject for each file, with filename as `name` and `digest` with the digest of the file contents.

I think we should leave the `Raw` subject alone, just like with the `image` exporter.

jedevc commented on exporter/attestation/extract.go:
```diff
@@ -0,0 +1,260 @@
+				if refs == nil {
+					return errors.Errorf("no refs map provided to lookup attestation keys")
+				}
+				ref, ok := refs[att.Ref]
```
Content doesn't make sense - for the bundle, the path is a directory path, not a file path. We iterate over the files in `unbundle`, so we'd need some weird way of representing the content here.

I've added an explicit error if `Content` is specified for the bundle.

jedevc commented on solver/result/attestation.go:
```diff
@@ -8,8 +8,9 @@ import (
-	Path string
+	Ref     string
+	Path    string
+	Content []byte
```
> Also, I'm not sure why you don't need conditions

Woops. It worked before because I was creating the `Content` attestations in `Unbundle` after this check. I've updated the checks here, and also made sure to add an explicit error in the gateway, if trying to send `Content` over it. Not sure about this design, depends on whether we use `Callback` instead of `Content`.

jedevc replied:
Have updated with `Callback` from https://github.com/moby/buildkit/pull/3240 - some minor differences in the conditions, for counting the number of attestations, etc, but should be functionally equivalent.

jedevc commented:
:bug: `StaticFS`/`MergeFS` have a bug with order of traversal, where `x.txt` will be sorted directly after `x/` because of the ASCII ordering - we need to apply a fix similar to in [contenthash](https://github.com/moby/buildkit/blob/master/cache/contenthash/checksum.go#L1178).

jedevc commented:
Rebased onto master after #3240 merged.

tonistiigi reviewed (commented):
Building with provenance attestation, I get a cryptic "context canceled" error. The issue seems to be that https://github.com/moby/buildkit/blob/f77133007f4b0044826169ec93c329713eb31261/solver/llbsolver/proc/provenance.go#L147 doesn't set a `Path`. In addition to fixing this, if this key is required it should be validated or at least the error should indicate what is going on.

jedevc commented:
Fixed the context cancelled issue. Also found a bug in trying to generate the hash for a file that is a symlink - I've refactored to only hash regular files, ignoring directories/symlinks/everything else.

jedevc commented on exporter/local/export.go:
```diff
@@ -148,7 +169,74 @@ func (e *localExporterInstance) Export(ctx context.Context, inp *exporter.Source
+						return errors.Wrap(err, "failed to marshal attestation")
+					}
+
+					if attestations[i].Path == "" {
```
Will resolve this one in a follow-up - agree that it should be somewhere else, but `AddAttestation` doesn't currently return an error, so might be a bit more invasive than the above comment.

jedevc commented on exporter/local/export.go:
```diff
@@ -148,7 +169,74 @@ func (e *localExporterInstance) Export(ctx context.Context, inp *exporter.Source
+				}
+				stmtFS := staticfs.NewFS()
+
+				now := time.Now().UnixNano()
```
Moved this out of the loop, but will keep the `Nano()` component for now - below we have `epoch.UnixNano()` so better to be consistent with that IMO.

tonistiigi replied:
> below we have epoch.UnixNano() so better to be consistent with that IMO.

This is only because the fsutil field expects nanoseconds. The actual `SOURCE_DATE_EPOCH` var is in full seconds.

tonistiigi commented on util/staticfs/static.go:
```diff
@@ -44,11 +45,11 @@ func (fs *FS) Add(p string, stat types.Stat, data []byte) {
 	for _, p := range keys {
-		st := fs.files[p].Stat
+		st := fs.files[convertKeyToPath(p)].Stat
 		if err := fn(p, &fsutil.StatInfo{Stat: &st}, nil); err != nil {
```
first parameter to `fn` should also be with `/`iiuc. 

Better to rename `s/p/k/` here as it is not the path anymore.

crazy-max reviewed (commented):
As discussed we should not set platform in output dir if only one is defined to be aligned with current local output logic. Prefix platform can be enforced for single platform using `BUILDKIT_MULTI_PLATFORM` attr: https://github.com/moby/buildkit/blob/efc63357b84c844d72d68e017dcdbe8741122380/frontend/dockerfile/builder/build.go#L451
